lib/backend/k8s/resources/customresource.go

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"reflect"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	extensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/rest"
+)
+
+// Interface required to satisfy use as a Kubernetes Custom Resource type.
+type CustomK8sResource interface {
+	runtime.Object
+	metav1.ObjectMetaAccessor
+}
+
+// Interface required to satisfy use as a Kubernetes Custom Resource List type.
+type CustomK8sList interface {
+	runtime.Object
+	metav1.ListMetaAccessor
+}
+
+// CustomK8sResourceConverter defines an interface to map between KVPair representation
+// and a custom Kubernetes resource.
+type CustomK8sResourceConverter interface {
+	// ListInterfaceToKey converts a ListInterface to a Key if the
+	// ListInterface specifies a specific instance, otherwise returns nil.
+	ListInterfaceToKey(model.ListInterface) model.Key
+
+	// Convert the Key to the Resource name.
+	KeyToName(model.Key) (string, error)
+
+	// Convert the Resource name to the Key.
+	NameToKey(string) (model.Key, error)
+
+	// Convert the Resource to a KVPair.
+	ToKVPair(CustomK8sResource) (*model.KVPair, error)
+
+	// Convert a KVPair to a Resource.
+	FromKVPair(*model.KVPair) (CustomK8sResource, error)
+}
+
+// customK8sResourceClient implements the K8sResourceClient interface and provides a generic
+// mechanism for a 1:1 mapping between a Calico Resource and an equivalent Kubernetes
+// custom resource type.
+type customK8sResourceClient struct {
+	clientSet       *kubernetes.Clientset
+	restClient      *rest.RESTClient
+	name            string
+	resource        string
+	description     string
+	k8sResourceType reflect.Type
+	k8sListType     reflect.Type
+	converter       CustomK8sResourceConverter
+}
+
+// newCustomK8sResourceClient creates a new customK8sResourceClient.
+func newCustomK8sResourceClient(c *kubernetes.Clientset, r *rest.RESTClient,
```

Comment:
Actually we could just drop the constructors since it just maps directly through to creating the struct.... so having an additional struct with the same data wouldn't make any sense here.